File: ts/test/abi-revision.test.ts
/**
 * The ABI-revision handshake, RUN against a fixture rather than assumed
 * (#36).
 *
 * docs/reference/artifact.md step 5 is normative: an artifact whose
 * `chs_abi_revision()` answers a value that is neither 0 nor the revision
 * this binding was written against MUST be rejected, naming BOTH numbers.
 * The fixture is a stub shared library built from the header itself by the
 * fixture generator the signed release serves as `abi-revision/gen.py`: one
 * registry root whose artifact answers `abi_revision + 1`
 * (`wrong-revision/`), one answering the header's own `abi_revision`
 * (`at-revision/`, the control — without it a binding whose own
 * `ABI_REVISION` had drifted would refuse everything and pass the refusal
 * case for entirely the wrong reason).
 *
 * Every expected number comes from the fixture's own `fixture.json`, never
 * from the exported `ABI_REVISION` — a case that read the constant under
 * test could not catch that constant being wrong. The refusal message is
 * checked with the fixture root's own text (and its realpath — macOS's
 * `/tmp` resolves to `/private/tmp`) stripped out first, so a digit that
 * happens to sit in a temp-directory name can never stand in for a digit the
 * binding itself was supposed to print.
 */

import { readFileSync, realpathSync } from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';
import { Registry, RegistryError } from '../src/index.js';

interface FixtureDoc {
  abi_revision: number;
  mismatch_revision: number;
  clickhouse_version: string;
  clickhouse_minor: string;
  platform?: string;
}

/**
 * The fixture root comes ONLY from this environment variable. There is no
 * fallback build here: a missing fixture means the suite is skipped LOUDLY
 * (the line below), never a silent pass.
 */
const ROOT = process.env['CHTYPES_ABI_FIXTURES'];
const HAVE_FIXTURES = ROOT !== undefined && ROOT !== '';

if (!HAVE_FIXTURES) {
  console.warn('[chtypes] abi-revision tests SKIPPED: no ABI revision fixture: $CHTYPES_ABI_FIXTURES is unset');
}

/**
 * Read and validate `fixture.json` eagerly, once, whenever a root was given.
 * A fixture that failed to generate correctly (missing, unparseable, or
 * internally inconsistent) must fail the suite loudly, naming the path —
 * never skip quietly and never let a case run against nonsense numbers.
 */
function loadFixture(root: string): FixtureDoc {
  const fixturePath = path.join(root, 'fixture.json');
  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(fixturePath, 'utf8'));
  } catch (err) {
    throw new Error(`chtypes: cannot read/parse ABI fixture at ${fixturePath}: ${String(err)}`);
  }
  const doc = parsed as Partial<FixtureDoc>;
  if (typeof doc.abi_revision !== 'number' || doc.abi_revision <= 0) {
    throw new Error(
      `chtypes: ABI fixture at ${fixturePath} has no positive abi_revision (got ${JSON.stringify(doc.abi_revision)})`,
    );
  }
  if (doc.mismatch_revision !== doc.abi_revision + 1) {
    throw new Error(
      `chtypes: ABI fixture at ${fixturePath} has mismatch_revision ${JSON.stringify(doc.mismatch_revision)}, ` +
        `expected abi_revision + 1 = ${doc.abi_revision + 1}`,
    );
  }
  return doc as FixtureDoc;
}

const DOC: FixtureDoc | null = ROOT !== undefined && ROOT !== '' ? loadFixture(ROOT) : null;

/**
 * "The message says N", not "the message contains the digits of N": a
 * refusal naming revision 4 must not be satisfied by the 4 in a path like
 * `.../24.8/`, and one naming 5 must not be satisfied by 15.
 */
function namesNumber(message: string, n: number): boolean {
  return new RegExp(`(^|[^0-9])${n}([^0-9]|$)`).test(message);
}

/**
 * Strip every occurrence of the fixture root — as given, and as
 * `fs.realpathSync` resolves it, since macOS's `/tmp` is a symlink to
 * `/private/tmp` and the two spellings can both appear depending on which
 * one a caller passed in — out of a message, longest first so the longer
 * (realpath) form is never left partially replaced by a shorter prefix
 * match. A digit sitting in a fixture directory name (a temp path, a run
 * id) must never be able to stand in for a digit the refusal itself is
 * required to print.
 */
function stripFixtureRoot(message: string, root: string): string {
  let real = root;
  try {
    real = realpathSync(root);
  } catch {
    // The root not resolving is not this helper's problem to raise; the
    // caller already has a live Registry error to report if anything here
    // was actually wrong.
  }
  const candidates = [...new Set([root, real])].sort((a, b) => b.length - a.length);
  let stripped = message;
  for (const candidate of candidates) {
    stripped = stripped.split(candidate).join('<fixture>');
  }
  return stripped;
}

describe.skipIf(!HAVE_FIXTURES)('abi revision fixture', () => {
  it('wrong revision is refused naming both numbers', () => {
    const doc = DOC!;
    const root = ROOT!;
    let thrown: unknown;
    try {
      new Registry(path.join(root, 'wrong-revision'));
    } catch (err) {
      thrown = err;
    }
    expect(thrown, 'a wrong-revision artifact must be refused at construction, not accepted').toBeInstanceOf(
      RegistryError,
    );
    const message = thrown instanceof Error ? thrown.message : String(thrown);
    const stripped = stripFixtureRoot(message, root);
    for (const n of [doc.mismatch_revision, doc.abi_revision]) {
      expect(
        namesNumber(stripped, n),
        `the refusal does not name ${n} (the artifact reports ${doc.mismatch_revision}, ` +
          `this binding speaks ${doc.abi_revision}): ${message}`,
      ).toBe(true);
    }
  });

  it('matching revision loads', () => {
    const doc = DOC!;
    const root = ROOT!;
    // Construction opens nothing, so the control artifact is opened by
    // `preload` — the constructor-time spelling of that request, and the one
    // whose failure is this test's verdict.
    const registry = new Registry(path.join(root, 'at-revision'), { preload: [doc.clickhouse_minor] });
    const libs = registry.libraries();
    expect(libs).toHaveLength(1);
    const lib = libs[0]!;
    expect(lib.abiRevision).toBe(doc.abi_revision);
    expect(lib.minor).toBe(doc.clickhouse_minor);
    expect(lib.version).toBe(doc.clickhouse_version);
    registry.close();
  });
});

File: go/chtypes/abi_revision_test.go
// abi_revision_test.go — the ABI-revision handshake, run against a
// wrong-revision fixture rather than assumed (#36).
//
// docs/reference/artifact.md step 5 is normative: an artifact whose
// chs_abi_revision() answers a value that is neither 0 nor the revision this
// binding was written against MUST be rejected, naming both numbers. This
// test drives that against two fixture registries the core repository's
// generator built from this repository's own include/chtypes.h: one artifact
// answering the header's revision plus one, one answering the header's own
// revision. Both expected numbers come from the fixture's fixture.json —
// which the generator wrote from the header — never from this package's own
// ABIRevision constant, because a case that read the constant under test
// could not catch the constant drifting.
package chtypes

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"testing"
)

// abiRevisionFixture is fixture.json, trimmed to the fields this test needs.
type abiRevisionFixture struct {
	Root              string
	ABIRevision       int    `json:"abi_revision"`
	MismatchRevision  int    `json:"mismatch_revision"`
	ClickHouseVersion string `json:"clickhouse_version"`
	ClickHouseMinor   string `json:"clickhouse_minor"`
	Platform          string `json:"platform"`
}

// loadABIRevisionFixture resolves the fixture set from $CHTYPES_ABI_FIXTURES
// alone. There is no fallback build and no other search: a CI census greps
// test output for the exact skip text below, so it must appear verbatim.
func loadABIRevisionFixture(t *testing.T) abiRevisionFixture {
	t.Helper()
	root := os.Getenv("CHTYPES_ABI_FIXTURES")
	if root == "" {
		t.Skip("no ABI revision fixture: $CHTYPES_ABI_FIXTURES is unset (build one with abi-revision/gen.py build --out DIR --header include/chtypes.h)")
	}
	raw, err := os.ReadFile(filepath.Join(root, "fixture.json"))
	if err != nil {
		t.Fatalf("ABI revision fixture %s: cannot read fixture.json: %v", root, err)
	}
	var doc abiRevisionFixture
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("ABI revision fixture %s: fixture.json does not parse: %v", root, err)
	}
	doc.Root = root
	if doc.ABIRevision <= 0 {
		t.Fatalf("ABI revision fixture %s: fixture.json reports abi_revision %d, want > 0", root, doc.ABIRevision)
	}
	if doc.MismatchRevision != doc.ABIRevision+1 {
		t.Fatalf("ABI revision fixture %s: fixture.json reports mismatch_revision %d, want abi_revision+1 (%d)",
			root, doc.MismatchRevision, doc.ABIRevision+1)
	}
	return doc
}

// namesNumber is "the message says N", not "the message contains the digits
// of N": a refusal naming revision 4 must not be satisfied by the 4 inside a
// path like .../24.8/, and one naming 5 must not be satisfied by 15.
func namesNumber(message string, n int) bool {
	return regexp.MustCompile(fmt.Sprintf(`(^|[^0-9])%d([^0-9]|$)`, n)).MatchString(message)
}

// scrubFixtureRoot removes every occurrence of the fixture root — and its
// symlink-resolved spelling, since macOS's /tmp vs /private/tmp differ — from
// message. Without this, namesNumber can be satisfied by a bare digit inside
// the fixture PATH itself (a scratchpad directory name, a request ID) rather
// than by anything the refusal actually said, which makes "names both
// numbers" vacuous on some machines. Longest spelling first, so one is never
// a no-op prefix of the other.
func scrubFixtureRoot(message, root string) string {
	forms := []string{root}
	if resolved, err := filepath.EvalSymlinks(root); err == nil && resolved != root {
		forms = append(forms, resolved)
	}
	sort.Slice(forms, func(i, j int) bool { return len(forms[i]) > len(forms[j]) })
	scrubbed := message
	for _, f := range forms {
		scrubbed = strings.ReplaceAll(scrubbed, f, "<fixture>")
	}
	return scrubbed
}

// TestABIRevisionMismatchIsRefused opens an artifact whose
// chs_abi_revision() answers the header's revision plus one. Refusal must
// happen at the call that OPENS it — not a Library that later degrades — and
// the error must name both the mismatched revision and this binding's own.
//
// Construction reads manifests and dlopens nothing, so the open is asked for
// with WithPreload: the constructor-time spelling of that request, and the one
// whose failure is this test's verdict.
func TestABIRevisionMismatchIsRefused(t *testing.T) {
	doc := loadABIRevisionFixture(t)
	wrong := filepath.Join(doc.Root, "wrong-revision")

	_, err := NewRegistry(wrong, WithPreload(doc.ClickHouseMinor))
	if err == nil {
		t.Fatalf("NewRegistry(%s, WithPreload(%q)) succeeded for an artifact reporting ABI revision %d; "+
			"docs/reference/artifact.md step 5 requires it be refused, naming both %d and %d",
			wrong, doc.ClickHouseMinor, doc.MismatchRevision, doc.MismatchRevision, doc.ABIRevision)
	}
	msg := err.Error()
	// Checked with the fixture root scrubbed out: a path segment can itself
	// hold a standalone digit (this scratchpad's own request ID does), so
	// checking the raw message could pass on the path alone rather than on
	// anything the refusal said. The failure text below still quotes the
	// original, unscrubbed message.
	scrubbed := scrubFixtureRoot(msg, doc.Root)
	if !namesNumber(scrubbed, doc.MismatchRevision) {
		t.Fatalf("refusal for %s does not name the mismatched revision %d: %s", wrong, doc.MismatchRevision, msg)
	}
	if !namesNumber(scrubbed, doc.ABIRevision) {
		t.Fatalf("refusal for %s does not name this package's own revision %d: %s", wrong, doc.ABIRevision, msg)
	}
}

// TestABIRevisionControlLoads is the negative control: the same stub, built
// at the header's own revision, must load cleanly. Without this, a binding
// that refused every artifact would pass the mismatch case above for
// entirely the wrong reason.
func TestABIRevisionControlLoads(t *testing.T) {
	doc := loadABIRevisionFixture(t)
	at := filepath.Join(doc.Root, "at-revision")

	// Preloaded for symmetry with the mismatch case above: both verdicts then
	// come from the same call on the same code path, which is the whole point
	// of a control.
	r, err := NewRegistry(at, WithPreload(doc.ClickHouseMinor))
	if err != nil {
		t.Fatalf("NewRegistry(%s, WithPreload(%q)) refused an artifact reporting the header's own ABI revision %d: %v",
			at, doc.ClickHouseMinor, doc.ABIRevision, err)
	}
	// Registry and Library are both documented as never needing a
	// close/shutdown (docs/reference/bindings.md §Teardown; multiversion.go's
	// Library doc comment) — there is no API to call here.

	lib, err := r.For(Version(doc.ClickHouseMinor))
	if err != nil {
		t.Fatalf("Registry.For(%q) on the control registry %s: %v", doc.ClickHouseMinor, at, err)
	}
	if lib.ABIRevision != doc.ABIRevision {
		t.Fatalf("control library at %s reports ABI revision %d, fixture.json says %d", at, lib.ABIRevision, doc.ABIRevision)
	}
	if string(lib.Version) != doc.ClickHouseVersion {
		t.Fatalf("control library at %s reports ClickHouse version %q, fixture.json says %q", at, lib.Version, doc.ClickHouseVersion)
	}
}
